codegen/smithy-codegen-core/src/main/java/software/amazon/smithy/codegen/core/ShapeIdShader.java

```diff
@@ -0,0 +1,255 @@
+
+        if (!ShapeId.isValidNamespace(target)) {
+            target = SANITIZE_CHARS.matcher(target).replaceAll("_");
+            target = SANITIZE_DOTS.matcher(target).replaceAll(".");
+            LOGGER.info(String.format("Sanitized %s `%s` to `%s`", TARGET_NAMESPACE, builder.targetNamespace, target));
+        }
+
+        rootNamespace = SmithyBuilder.requiredState(ROOT_NAMESPACE, builder.rootNamespace);
+        targetNamespace = target;
+    }
+
+    /**
+     * Builds a new {@code ShapeIdShader}.
+     *
+     * @return Returns the builder.
+     */
+    public static Builder builder() {
+        return new Builder();
+    }
+
+    /**
+     * Creates a shaded version of the given shape ID.
+     *
+     * @param id Shape ID to shade.
+     * @return Returns the shaded shape ID.
+     */
+    public ShapeId shade(ShapeId id) {
+        return shade(id, null);
+    }
+
+    /**
+     * Creates a shaded version of the given shape ID and combines
+     * any additional namespace segments using the provided
+     * {@code ShadeOption} strategy.
+     *
+     * @param id Shape ID to shade.
+     * @param option The shade strategy to use.
+     * @return Returns the shaded shape ID.
+     */
+    public ShapeId shade(ShapeId id, ShadeOption option) {
+        String name = id.getName();
+        String namespace = shadeNamespace(id);
+
+        if (namespace.isEmpty()) {
+            namespace = targetNamespace;
+        } else if (Objects.equals(option, ShadeOption.SQUASH_INTO_NAMESPACE)) {
+            namespace = targetNamespace + "." + squashName(namespace);
+        } else if (Objects.equals(option, ShadeOption.SQUASH_INTO_NAME)) {
+            name = squashName(namespace + "." + name);
+            namespace = targetNamespace;
+        } else {
+            namespace = targetNamespace + "." + namespace;
+        }
+
+        if (namespace.equals(id.getNamespace()) && name.equals(id.getName())) {
+            return id;
+        }
+
+        return ShapeId.fromParts(namespace, name, id.getMember().orElse(null));
+    }
+
+    private String squashName(String name) {
+        StringBuilder builder = new StringBuilder();
+        for (String identifier : DOT.split(name)) {
+            builder.append(StringUtils.capitalize(identifier));
+        }
+        return builder.toString();
+    }
+
+    private String shadeNamespace(ShapeId id) {
+        String namespace = id.getNamespace();
+
+        if (namespace.equals(rootNamespace)) {
+            return "";
+        }
+
+        if (namespace.startsWith(rootNamespace + ".")) {
+            // Strip out the root namespace.
+            namespace = namespace.substring(rootNamespace.length() + 1);
+        }
+
+        return namespace;
+    }
+
+    /**
+     * Controls how shading is handled when there are namespace segments found
+     * outside of the {@code rootNamespace}.
+     */
+    public enum ShadeOption {
```

Comment:
This should probably have an `APPEND` (or similar) that represents the default behavior.